### PR TITLE
WIP: Improve coverage

### PR DESCRIPTION
**Pull request by allevo**
This PR aims to raise the code coverage (https://github.com/fastify/fastify/issues/289)

Some troubles occurred:
- [those tests](https://github.com/allevo/fastify/blob/e6c2b7882ac8b82c649ab1a7223eb6e45fe7f885/test/middleware.test.js#L484) should be reviewed and fixed before merging: the behaviour is unexpected.
- IMHO [this](https://github.com/allevo/fastify/blob/e6c2b7882ac8b82c649ab1a7223eb6e45fe7f885/test/throw.test.js#L124) and [this](https://github.com/allevo/fastify/blob/e6c2b7882ac8b82c649ab1a7223eb6e45fe7f885/test/throw.test.js#L139) are bugs
- I don't know where (in which test file) we should test [those lines](https://github.com/fastify/fastify/blob/master/lib/handleRequest.js#L48)
- I don't know how to test [this](https://github.com/fastify/fastify/blob/master/lib/handleRequest.js#L57)
- `lib/decorate.js` file needs more test

### CODE REVIEW

**Comment by StarpTech** [comment]
+1

> I don't know how to test this

From node documentation

> If any error is encountered during the request (be that with DNS resolution, TCP level errors, or actual HTTP parse errors) an 'error' event is emitted on the returned request object. As with all 'error' events, if no listeners are registered the error will be thrown.

Try to send an invalid http request? :/

**Comment by allevo** [comment]
@StarpTech no:
DNS resolution is done before the request arrive to the server. The HTTP parse error is not possible there because node has already parsed the http request (at least the header).
Maybe with a dropped chunked request... I need to figure out and make some tests...

**Comment by mcollina** [comment]
I think decorate is fine as it is, so a plugin could specify an “abstract” implementation that throws an exception. I think we should discuss this, becuase it’s a grey area of the API. I’m ok in throwing.

**Comment by mcollina** [comment]
the behavior of the middlewares is fixed in the 404 PR.

**Comment by mcollina** [comment]
@allevo you should rebase on top of master.

How much this improve the code coverage?

@delvedor you should check out why coveralls is not reporting the code coverage to the PRs anymore (sigh), I don't know why I can't access the settings myself.

**Comment by allevo** [comment]
@mcollina After the rebase, [those tests](https://github.com/allevo/fastify/blob/e6c2b7882ac8b82c649ab1a7223eb6e45fe7f885/test/middleware.test.js#L484) fail again. [this](https://github.com/allevo/fastify/blob/e6c2b7882ac8b82c649ab1a7223eb6e45fe7f885/test/throw.test.js#L124) and [this](https://github.com/allevo/fastify/blob/e6c2b7882ac8b82c649ab1a7223eb6e45fe7f885/test/throw.test.js#L139) too.

I uncomment all "TODO" that are strange. Maybe they are bugs, maybe I understood wrongly but at least they are visible!

**Comment by allevo** [comment]
Final repo:

```
-----------------------|----------|----------|----------|----------|----------------|
File                   |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------------|----------|----------|----------|----------|----------------|
All files              |    98.35 |     96.9 |    98.35 |    98.47 |                |
 fastify               |    98.01 |    95.83 |    97.96 |    98.31 |                |
  fastify.js           |    98.01 |    95.83 |    97.96 |    98.31 |... 552,553,554 |
 fastify/lib           |    98.63 |    97.77 |    98.61 |     98.6 |                |
  ContentTypeParser.js |      100 |      100 |      100 |      100 |                |
  decorate.js          |    94.12 |       95 |      100 |    94.12 |          30,31 |
  handleRequest.js     |    98.53 |      100 |    92.86 |    98.51 |             58 |
  hooks.js             |      100 |      100 |      100 |      100 |                |
  logger.js            |    93.75 |    83.33 |      100 |    93.75 |          76,81 |
  reply.js             |      100 |    95.83 |      100 |      100 |          67,78 |
  request.js           |      100 |      100 |      100 |      100 |                |
  validation.js        |      100 |      100 |      100 |      100 |                |
-----------------------|----------|----------|----------|----------|----------------|
```

The missing lines are:
- `fourOhFourFallBack` in fastify.js where there's written that it should be impossible to land there (????)
- `checkDependencies` in logger.js where I cannot understand how to enter in the "if"
- the error case in `jsonBody` in handleRequest.js where I cannot find a way to emit an error on the request object
- `onResponseCallback` in logger where I didn't understand how to emit error on response object

Anyway, this is the best I could do with my knowledge. If there're suggestions, they are welcomed!

**NB**: the test of the prefixed middleware still fail but I need to know if the test expectations are right before fixing the code!

The failed test is [here](https://github.com/allevo/fastify/blob/improve-coverage/test/middleware.test.js#L504)

**Comment by mcollina** [comment]
`fourOhFourFallBack` should stay not covered, it's there as a safeguard.

I don't think we are supporting `'/'` and `''` as prefixes for middlewares (this is a bug). You should probably include a fix, good spot. The fix might go into middie.

**Comment by allevo** [comment]
@mcollina I'll open PR to middie if I manage to re-propose the bug.

File: lib/decorate.js [hunk 1]
```diff
@@ -34,7 +38,7 @@ function checkDependencies (instance, deps) {
 }
 
 function decorateReply (name, fn, dependencies) {
-  if (checkExistence(this._Reply, name)) {
+  if (checkExistenceInPrototype(this._Reply, name)) {
```
Comment (mcollina) [inline lib/decorate.js]: let's just do `name in this._Reply`

File: test/405.test.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+  const fastify = Fastify()
+
+  fastify.get('/', function (req, reply) {
+    reply.send({ hello: 'world' })
+  })
+
+  t.tearDown(fastify.close.bind(fastify))
+
+  fastify.listen(0, err => {
```
Comment (mcollina) [inline test/405.test.js]: can we use `inject` instead? It will end up faster.

File: lib/decorate.js (hunk 1 above)
Comment (allevo) [reply lib/decorate.js]: `name` is not in `this._Reply`. `name` is in its prototype.

If you want we could write `name in this._Reply.prototype`

**Comment by allevo** [comment]
After rebase:
```
-----------------------|----------|----------|----------|----------|----------------|
File                   |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-----------------------|----------|----------|----------|----------|----------------|
All files              |    98.35 |     96.9 |    98.36 |    98.47 |                |
 fastify               |    98.01 |    97.22 |       98 |    98.32 |                |
  fastify.js           |    98.01 |    97.22 |       98 |    98.32 |... 558,559,560 |
 fastify/lib           |    98.63 |    96.65 |    98.61 |     98.6 |                |
  ContentTypeParser.js |      100 |      100 |      100 |      100 |                |
  decorate.js          |    94.12 |       95 |      100 |    94.12 |          30,31 |
  handleRequest.js     |    98.53 |    95.45 |    92.86 |    98.51 |             58 |
  hooks.js             |      100 |      100 |      100 |      100 |                |
  logger.js            |    93.75 |    83.33 |      100 |    93.75 |          76,81 |
  reply.js             |      100 |    95.83 |      100 |      100 |          67,78 |
  request.js           |      100 |      100 |      100 |      100 |                |
  validation.js        |      100 |      100 |      100 |      100 |                |
-----------------------|----------|----------|----------|----------|----------------|
```

**Comment by mcollina** [comment]
Tests are still failing, is it still expected?

**Comment by allevo** [comment]
yeah, I'm waiting for this https://github.com/fastify/middie/pull/12.
But maybe we can remove those test because it checks if `middie` works properly

**Comment by mcollina** [comment]
Landing now, it's ok! :)

**Pull request merged**